### ISSUE TITLE
AnimComposer: Add Features/Potential issues

### ISSUE THREAD

**Issue by TehLeo** [jMonkeyEngine/jmonkeyengine#1079]
To be examined and confirmed:
- tested animation "Standard Walk" from mixamo, the animation does not loop smoothly -> there is a jump at the point when the animation should repeat. (tried deleting the last frame, etc, it did not fix the issue)
-to test the issue, I set the speed of layer to zero and on key press set speed to one, manually call update, set speed to zero, in order to jump from frame to frame. To my surprise, this produced fluid motion at the point where the animation repeats.  

potential new features
-loop modes (cannot run animation exactly once) / listeners
-set animation time directly/set frame directly

**Comment by Ali-RS** [comment]
I added support for looping in this PR: #1038

**Comment by tonihele** [comment]
Yeah, I'm trying to convert our project to use the new animation system. The looping issue is a major thing for us. I understand the workaround and tried it. But as @Ali-RS pointed out, running the animation backwards doesn't work since all control is lost over the looping.

To be more specific, I tried to to re-create the ye olden loop modes in our code by creating a custom action. Cycle mode would then set the animation speed -1*speed when animation has reached the end. And it stops triggering after the speed is set to negative.

What was wrong with PR #1038 ? Wouldn't it have solved this?

**Comment by Ali-RS** [comment]
> What was wrong with PR #1038 ? 

I think it had not been reviewed. I am not sure why I closed it though!

**Comment by Ali-RS** [comment]
Regarding

> there is a jump at the point when the animation should repeat.

Maybe `BlendableAction.setTransitionLength(0)` can help with this. See https://hub.jmonkeyengine.org/t/solved-animation-does-not-loop-properly/43272/8?u=ali_rs

**Comment by tonihele** [comment]
> 
> 
> > What was wrong with PR #1038 ?
> 
> I think it had not been reviewed. I am not sure why I closed it though!

Yes, since I understood from the discussion that it is just waiting for the review :) Ideally looping, like in the old, Cycle, Loop, NoLooping would be built in. But if at least the bug is fixed with the backwards animation, I would be really happy. That could affect other stuff as well, and really, is a bug I think.

**Comment by Ali-RS** [comment]
In my case, I do not use AnimComposer for playing tweens, instead, I use TweenAnimation. There I can also control looping.

https://github.com/jMonkeyEngine-Contributions/Lemur/blob/3448a1621bdc7c99be9aeb7d8061da5babd62705/src/main/java/com/simsilica/lemur/anim/TweenAnimation.java#L48

and run them with AnimationState

https://github.com/jMonkeyEngine-Contributions/Lemur/blob/3448a1621bdc7c99be9aeb7d8061da5babd62705/src/main/java/com/simsilica/lemur/anim/AnimationState.java#L51

For looping, I think the best way is to wrap it inside a Loop tween (afaik we do not have one in JME's stock tweens so you need to create it yourself).

https://hub.jmonkeyengine.org/t/tweens-smoothstep-not-always-smooth-oto-mesh-xml/43353/23?u=ali_rs

For backward playing, I simply wrap it inside an inverse tween.

**Comment by tonihele** [comment]
How would you construct this reverse tween? I tried to make this action thingie to control the animation direction... https://github.com/tonihele/OpenKeeper/blob/18a04d03e361b2cbdea80c3b2a48edae5891952f/src/toniarts/openkeeper/view/animation/LoopAction.java

**Comment by Ali-RS** [comment]
>How would you construct this reverse tween?


Something like this: 

    public class Invert implements Tween {
    
        private Tween delegate;
        
        public Invert( Interpolator delegate ) {
            this.delegate = delegate;
        }
         
        public void interpolate( double t ) {
            delegate.interpolate(delegate.getLenght() - t);
        }
        
        @Override
        public String toString() {
            return "Invert[delegate=" + delegate + "]";  
        }
    }
